src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/BatchEmitter.java

```diff
@@ -276,6 +302,7 @@ public boolean add(final TrackerPayload payload) {
         
         if (!result) {
             LOGGER.error("Unable to add payload to emitter, emitter buffer is full");
+            callback.onFailure(FailureType.TRACKER_STORAGE_FULL, false, Collections.singletonList(payload));
```

Comment:
This may be just my personal preference, but I find it a bit confusing that we call the callback for the `TRACKER_STORAGE_FULL` failure type in two different classes (BatchEmitter and InMemoryStorage). I think it'd be better to define who is responsible for handling the failure type and handle all its instances in that class. If we can't handle all callbacks in the Emitter (which I think would be clearer than delegating the responsibility to other places), could we handle all `TRACKER_STORAGE_FULL` errors in the InMemoryStorage?

Reply:
Nice idea!